drivers/SAI.cpp

```diff
@@ -0,0 +1,71 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2013 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "drivers/SAI.h"
+#include "platform/mbed_critical.h"
+
+#if DEVICE_SAI
+
+namespace mbed {
+
+SAI::SAI(const sai_format_t *fmt, bool is_input, uint32_t master_clock, bool internal_mclk) : _sai(), _mutex(), _is_input(is_input) {
```

Comment:
I have some doubts regarding the SAI API. I think that API should provide more 
straight forward settings, so the user is able to easily configure SAI according to his needs. Currently we have complicated `sai_format_t` based on which the driver should deduct the required configuration (e.g. used protocol based on `ws_length` field of given format). I suggest to create config structure with typical SAI settings like:
- Operating mode (master os slave),
- Format of frames/protocol (I2S, MSB Justified, LSB Justified, PCM Short Frame, PCM Long Frame),
- Used channels (stereo, left, right),
- Width and alignment of samples (`sai_format_t` might be reduced and provide only this info),
- Sampling frequency,
- Sync modes (synchronous, asynchronous),
- Clock settings (We got `master_clock` which defines freq and `internal_mclk` which defines if internal or external master clock is used. We might add Enable Master Clock output flag).

I think that this would be a good start. We can expand the setting if needed later.
`sai_init()` function should check if the specific device is able to handle provided config and return initialisation status. Since this is done in constructor additional member function should be provided to check if SAI object has been successfully initialised.

Additionally we may add member function which returns hardware capabilities (e.g. `SAI_GetCapabilities()`).

Reply:
Indeed,

The C++ API is only provisioned here for the sake of running some reference implementation demo app.
It is not by any mean a final API to present to users.

The typical format I proposed here are 
```c
/** SAI configuration for I2S Philips 16 bit data & word size */
extern const sai_format_t sai_mode_i2s16;
/** SAI configuration for I2S Philips 16bit data & 32bit word size */
extern const sai_format_t sai_mode_i2s16w32;
/** SAI configuration for I2S Philips 32 bit data & word size */
extern const sai_format_t sai_mode_i2s32;
/** SAI configuration for PCM 16 bit data & word size with long sync */
extern const sai_format_t sai_mode_pcm16l;
/** SAI configuration for PCM 16 bit data & word size with short sync */
extern const sai_format_t sai_mode_pcm16s;
```

SoCs having this API should support at least all of these formats.

In regards to clocks supports they vary a lot between platforms (eg. some may or may not support custom oversampling).
As it is highly dependent on the hardware, it probably make not sense to have predefined clock configurations.

Reply:
The thing you quote here is very restrictive compared to what some HW can achieve.

Let's imagine some application having a 7 microphones array.
Your ADCs requires you to receive :
 - words : 15bit lsb first
 - data  :
	- 12bit right aligned data
 - BClock: Low polarity
 - fsync :
	- 4 bit long starting 1 bit before the start of frame 
	- High polarity

The ADCs are time multiplexed on a period of 16 words.
Your system is neither mono nor stereo, it is microphone array.

PCM/I²S doesn't make any sense on their own in this context but this is still a valid and supported SAI configuration (supported by some platforms).

We do not aim at defining all possible configurations with these constants, but only provide configurations that MUST be supported by SoC to be tagged as having SAI.
They may support more configurations than these but not less.

In regards to the clock, the field naming might not be the best yet but you have 4 main parameters :
 - the Sample Rate
 - the Clock Source :
	- Internal (using peripheral clock)
	- External (using MCLK pin)
	- Sibling  (when configuring the receiver, using the transmitter clock, could it be it-self internal or external)
 - the input MCLK clock frequency
 - the output MCLK clock frequency
 
The last two are because you might receive an external MCLK coming from your sibling SAI & re-emit a MCLK with another frequency.
 
hm, by writing this, I see how I could reorganise these parameters:
```c
typedef struct sai_clock_source_e {
	SAI_CLOCK_SOURCE_Internal,
	SAI_CLOCK_SOURCE_External,
	SAI_CLOCK_SOURCE_Sibling,
} sai_clock_source_t;

typedef struct sai_init_s {
	...
	sai_clock_source_t mclk_source;
	uint32_t 	input_mclk_frequency;
	uint32_t 	output_mclk_frequency;
	uint32_t	sample_rate;
	...
} sai_init_t;
```